c10/util/typeid.h

```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
 #endif
   }
 
- private:
-  const detail::TypeMetaData* data_;
+  /**
+  * convert ScalarType enum values to TypeMeta handles
+  */
+  static inline caffe2::TypeMeta fromScalarType(ScalarType scalar_type) {
+    const size_t index = static_cast<uint16_t>(scalar_type);
+    if (C10_LIKELY(index < NumScalarTypes)) {
+      return TypeMeta(index);
+    }
+    error_unrecognized_scalartype(scalar_type);
+    return TypeMeta(); // avoids "control reaches end of non-void function"
+  }
 
+  /**
+   * convert TypeMeta handles to ScalarType enum values
+   */
+  inline ScalarType toScalarType() {
+    if (C10_LIKELY(isScalarType())) {
+      return static_cast<ScalarType>(index_);
+    }
+    error_unsupported_typemeta(*this);
+    return ScalarType::Undefined; // avoids "control reaches end of non-void function"
+  }
+
+private:
+  static void error_unrecognized_scalartype(ScalarType scalar_type) {
+    TORCH_CHECK(false, "Unrecognized Scalartype ", scalar_type, " (please report this error)");
+  }
+  static void error_unsupported_typemeta(caffe2::TypeMeta dtype) {
+    TORCH_CHECK(false, "Unsupported TypeMeta in ATen: ", dtype, " (please report this error)");
+  }
+
+  // preallocated TypeMetaData instances for ScalarType types
+  static const detail::TypeMetaData& scalarTypeMetaData(uint16_t index);
+
+  // TypeMetaData instances for non-ScalarType types...
+  static std::vector<detail::TypeMetaData>& nonScalarTypeMetaDatas();
+
+  // ...added through here
+  static std::mutex appendMutex_;
   template <class T>
-  C10_API static const detail::TypeMetaData* _typeMetaDataInstance() noexcept;
+  static uint16_t addTypeMetaDataInstance() {
+    auto typeId = TypeIdentifier::Get<T>();
+    auto typeName = c10::util::get_fully_qualified_type_name<T>();
+    std::vector<detail::TypeMetaData>& instances = nonScalarTypeMetaDatas();
+    // concurrency situation:
+    // the static init for each _typeMetaDataInstance<T> is synchronized,
+    // so we don't need to worry about the same type being added twice.
+    // But different threads might be adding different Ts concurrently.
+    // So while we don't need to test for the presence of T before
+    // appending, we do need to serialize the appends themselves.
+    std::lock_guard<std::mutex> lock(appendMutex_);
+    uint16_t size = instances.size();
+    instances.emplace_back(
+      sizeof(T),
+      detail::_PickNew<T>(),
+      detail::_PickPlacementNew<T>(),
+      detail::_PickCopy<T>(),
+      detail::_PickPlacementDelete<T>(),
+      detail::_PickDelete<T>(),
+      typeId,
+      typeName);
+    return size + NumScalarTypes;
+  }
+
+  // specializations return indexes into typeMetaDataInstances()
+  template <class T>
+  C10_API static uint16_t _typeMetaDataInstance() noexcept;
+
+  //
+  // TypeMeta just wraps this index
+  //
+
+  uint16_t index_;
+
+  inline const detail::TypeMetaData& data() const {
```

Comment:
The if statement here seems totally unnecessary. Why not just initialize `nonScalarTypeMetaDatas` to start with the scalar type metadatas? Then you don't need to if statement here at all.

Reply:
I was typing the same comment then seeing yours :)

Reply:
You mean just pad it out? Sure why not. They wouldn't even need to be the actual scalar type metadatas, just the equivalent number of blanks - that would prevent people from accessing the dynamic vector instead of the static array.

Reply:
The original motivation was to save the extra hop to the heap-allocated (and volatile) vector backing store, and also take advantage of the array being constexpr, back when I thought it could be made constexpr unconditionally. The latter didn't pan out, but the former was paying off pretty nicely - until the fast-path `itemsize()` took the heat off this call path. The current savings on the microbenchmarks are tiny, and it's hard to see this path heating up on other use cases, so probably not worth the complexity. Assuming the solution to the concurrency problem doesn't slow down the dynamic vector to the point where the payoff here again becomes interesting, I'll get rid of the static one.

Reply:
I mean, if the vector is going to have fixed max size, you can just preallocate a big array and then you get the best of both worlds (since we can save the hop).